### PR TITLE
Mission edit using its name

### PR DESCRIPTION
**Pull request by Smieszkokoleszko**
New feature to start editing mission using its name instead of index.

Able to merge after PR #31

### CODE REVIEW

**Comment by 3Mydlo3** [comment]
Converted to draft as it fails due to dependency on #31

**Comment by 3Mydlo3** [comment]
Build failed with:
```
D:\a\Boderator\Boderator\ArmaforcesMissionBot\Modules\Signups.cs(666,54): error CS1061: 'SignupsData' does not contain a definition for 'GetCurrentlyEditedMission' and no accessible extension method 'GetCurrentlyEditedMission' accepting a first argument of type 'SignupsData' could be found (are you missing a using directive or an assembly reference?) [D:\a\Boderator\Boderator\ArmaforcesMissionBot\ArmaforcesMissionBot.csproj]
```
@Smieszkokoleszko Please fix this and mark as ready for review when all works.

**Comment by Smieszkokoleszko** [comment]
I know about the issue sometimes I'm just busy. Added a file. It has been already added in PR #31. Checks passed.

File: ArmaforcesMissionBot/Modules/Signups.cs [hunk 1]
```diff
@@ -657,6 +657,40 @@ public async Task EditMission(int missionNo)
...
+        [Command("edytuj-misje")]
+        [Summary("Po podaniu nazwy misji jako parametru włączy edycje danej misji (część bez zespołów).")]
+        [ContextDMOrChannel]
+        public async Task EditMission([Remainder] string title)
+        {
+            var currentlyEditedMission = SignupsData.GetCurrentlyEditedMission(Context.User.Id);
+
+            if (currentlyEditedMission == null)
+            {
+                var missionToBeEdited = SignupsData.Missions.FirstOrDefault(x => x.Title == title);
```
Comment (3Mydlo3) [inline ArmaforcesMissionBot/Modules/Signups.cs]: It doesn't always work if you call the command `AF!edytuj-misje mission-title`. The channel name could be `#mission-title` but the mission name is e.g. `Mission Title` and it might be confusing so the comparison should be done some other way if we allow to specify the string instead of channel "mention".

File: ArmaforcesMissionBot/Modules/Signups.cs [hunk 2]
```diff
@@ -657,6 +657,40 @@ public async Task EditMission(int missionNo)
                 }
             }
         }
+		
+        [Command("edytuj-misje")]
+        [Summary("Po podaniu nazwy misji jako parametru włączy edycje danej misji (część bez zespołów).")]
+        [ContextDMOrChannel]
+        public async Task EditMission([Remainder] string title)
```
Comment (3Mydlo3) [inline ArmaforcesMissionBot/Modules/Signups.cs]: I think it would be much better to create this overload:
```suggestion
        public async Task EditMission(IGuildChannel channel)
```
It will be much easier to use.

No need for the `string` overload if you don't want to allow specifying only part of the mission name e.g. `AF!edytuj-misje Mission Tit` for the mission titled `Mission Title`. I guess that's not needed with channel mentions and Discord suggestions in place.

File: ArmaforcesMissionBot/Modules/Signups.cs (hunk 1 above)
Comment (3Mydlo3) [reply ArmaforcesMissionBot/Modules/Signups.cs]: I think you should reuse the channel name creation from mission title. Get the mission title, create a channel name for it and then compare the strings. When using the `IGuildChannel` like suggested above, these 2 should always equal.

File: ArmaforcesMissionBot/Modules/Signups.cs [hunk 3]
```diff
@@ -657,6 +657,40 @@ public async Task EditMission(int missionNo)
...
+                if (missionToBeEdited.Owner != Context.User.Id)
+                {
+                    await ReplyAsync($"Nie nauczyli żeby nie ruszać nie swojego?");
+                    return;
+                }
+
+                // Don't want to write another function just to copy class, and performance isn't a problem here so just serialize it and deserialize
+                var serialized = JsonConvert.SerializeObject(missionToBeEdited);
+                SignupsData.BeforeEditMissions[Context.User.Id] = JsonConvert.DeserializeObject<ArmaforcesMissionBotSharedClasses.Mission>(serialized);
+                missionToBeEdited.Editing = ArmaforcesMissionBotSharedClasses.Mission.EditEnum.Started;
```
Comment (3Mydlo3) [inline ArmaforcesMissionBot/Modules/Signups.cs]: I guess this should be extracted somewhere because it's a copy paste from other overload. No need to extract if you decide to remove the index based mission edition.

File: ArmaforcesMissionBot/Modules/Signups.cs [hunk 4]
```diff
@@ -657,6 +657,40 @@ public async Task EditMission(int missionNo)
                 }
             }
         }
```
Comment (3Mydlo3) [inline ArmaforcesMissionBot/Modules/Signups.cs]: I'd remove the index based edition as it's totally useless when we can just mention the channel.

File: ArmaforcesMissionBot/Extensions/SignupsDataExtensions.cs [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+    {
+        public static Mission GetCurrentlyEditedMission(this SignupsData signupsData, ulong userId)
+        {
+            return signupsData.Missions.SingleOrDefault(x =>
+                x.Owner == userId &&
+                ((x.Editing == ArmaforcesMissionBotSharedClasses.Mission.EditEnum.New) ||
+                (x.Editing == ArmaforcesMissionBotSharedClasses.Mission.EditEnum.Started)));
+        }
+    }
+}
```
Comment (3Mydlo3) [inline ArmaforcesMissionBot/Extensions/SignupsDataExtensions.cs]: Next time please merge the master to feature branch instead of adding the changes yourself.

**Review by 3Mydlo3 — CHANGES_REQUESTED** [review]

File: ArmaforcesMissionBot/Modules/Signups.cs (hunk 1 above)
Comment (Smieszkokoleszko) [reply ArmaforcesMissionBot/Modules/Signups.cs]: But why? The descriptions states clearly that the argument is mission name. Not channel name or anything else. Making this comparison case insensitive is the only thing I see reasonable.  User writes a name of the mission like it is everywhere except channel name so why using channel name?

Comment (3Mydlo3) [reply ArmaforcesMissionBot/Modules/Signups.cs]: What's the point of copy-pasting whole mission title (because you will need to enter the mission channel, copy the title and come back to input a command)? Why not just use the `IGuildChannel` and use the channel mention which is the only intuitive way of invoking commands like this? Just like for `AF!ban` you would need to enter the user name instead of just mentioning him.

Comment (Smieszkokoleszko) [reply ArmaforcesMissionBot/Modules/Signups.cs]: OK. Now I get it and it makes sense. I thought about writing channel name by hand.

**Comment by Smieszkokoleszko** [comment]
Ready to review.

File: ArmaforcesMissionBot/Modules/Signups.cs [hunk 6]
```diff
@@ -619,20 +619,27 @@ public async Task CancelMission(int missionNo)
...
+                
+                if (missionToBeEdited.Owner != Context.User.Id)
+                {
+                    await ReplyAsync($"Nie nauczyli żeby nie ruszać nie swojego?");
+                    return;
+                }
 
-                // Don't want to write another function just to copy class, and performance isn't a problem here so just serialize it and deserialize
-                var serialized = JsonConvert.SerializeObject(missionToBeEdited);
-                SignupsData.BeforeEditMissions[Context.User.Id] = JsonConvert.DeserializeObject<ArmaforcesMissionBotSharedClasses.Mission>(serialized);
```
Comment (3Mydlo3) [inline ArmaforcesMissionBot/Modules/Signups.cs]: Why was this removed? Did you investigate it does nothing and can be safely removed or it was just an accident? If it doesn't break anything then that's great, two useless lines less.

I'm really curious if the `SignupsData.BeforeEditMissions` is used anywhere or is just some legacy code.

**Review by 3Mydlo3 — APPROVED** [review]

File: ArmaforcesMissionBot/Modules/Signups.cs (hunk 6 above)
Comment (Smieszkokoleszko) [reply ArmaforcesMissionBot/Modules/Signups.cs]: My mistake as I over thought that. Code used for saving mission state before editing in case of cancelling. Done in this way to have pure data, not a reference.
